src/clj_pgp/core.clj

```diff
@@ -262,21 +262,36 @@
   (.nextObject factory))
 
 
+(defn ^:no-doc try-read-object
+  "Tries to read and return a single object from a given
+  PGPObjectFactory calling the error/*handler* upon an exception."
+  [factory input n]
+  (try
+    (read-next-object factory)
+    (catch Exception e
+      (error/*handler* ::read-object-error
+                       (.getMessage e)
+                       (assoc (ex-data e) ::stream input ::nth n)
+                       e))))
+
+
 (defn ^:no-doc read-objects
-  "Lazily decodes a sequence of PGP objects from an input stream."
+  "Decodes a sequence of PGP objects from an input stream."
   [^InputStream input]
-  (let [factory (PGPObjectFactory. input (BcKeyFingerprintCalculator.))]
-    (->> (range)
-         (map (fn next-object
-                [n]
-                (try
-                  (read-next-object factory)
-                  (catch Exception e
-                    (error/*handler* ::read-object-error
-                                     (.getMessage e)
-                                     (assoc (ex-data e) ::stream input ::nth n)
-                                     e)))))
-         (take-while some?))))
+  (reify
+    clojure.lang.IReduceInit
+    (reduce
+      [_ rf init]
+      (let [factory (PGPObjectFactory. input (BcKeyFingerprintCalculator.))]
+        (loop [acc init
+               n 0
+               obj (try-read-object factory input n)]
```

Comment:
Why is the object included in the loop here? It forces you to repeat this `try-read-object` call in the `recur`, but you could read it once in a `let` at the start of the loop instead.